[로또] 큰곰 미션 제출합니다. (3단계까지)

kunggom opened the pull request:
추후 웹 버전 구현 이전에, 먼저 콘솔 버전의 미션을 제출합니다.

kunggom committed:
추가: 미션 3단계까지 완료

gracefulBrown commented on src/main/java/lotto/model/LottoRule.java:
```diff
@@ -0,0 +1,75 @@
+        this.numberEnd = end;
+        this.numberCount = count;
+        this.price = price;
+        this.numberFormatMessage
```
출력과 관련된 로직은 출력뷰로 옮기는건 어떨까요?

kunggom replied:
`this.numberFormatMessage` 말씀하시는 거지요? 이 부분은 객체 자체에서 예외 메시지를 만들기 위한 부분인데, 그걸 출력 뷰로 옮기는 게 맞는 건가요?

gracefulBrown replied:
에러 메시지라면 출력과 관계된 로직이 아니군요 :) 무시하셔도 됩니다 :)

gracefulBrown commented on src/main/java/lotto/model/LottoRule.java:
```diff
@@ -0,0 +1,75 @@
+                + "개로 이루어져 있어야 합니다.";
+    }
+
+    public int getNumberStart() {
```
사용하지 않는 getter는 제거해주셔도 좋습니다 :)

kunggom replied:
차후 미션에서 사용할 것 같아서 미리 만들어둔 부분인데, 주석 처리하도록 하겠습니다.

gracefulBrown commented on src/main/java/lotto/controller/LottoGame.java:
```diff
@@ -0,0 +1,110 @@
+import java.util.ArrayList;
+import java.util.List;
+
+class LottoGame {
```
해당 클래스는 사용하고 있지 않는건가요?

kunggom replied:
위와 마찬가지로, 3단계 이후의 미션 진행용으로 작성한 것입니다. 콘솔 버전은 콘솔 입출력에 너무 묶여 있는 것 같아 별도로 작성하였는데, 웹 버전에서 사용할 예정입니다.

gracefulBrown commented on src/main/java/lotto/view/Input.java:
```diff
@@ -0,0 +1,64 @@
+    public static List<Integer> tryGetNumbers() {
+        Stream<String> temp = Arrays.stream(getInput(EMPTY));
+        try {
+            return temp.map(Integer::parseInt).collect(Collectors.toList());
```
스트림 메서드 호출은 한줄에 하나씩 하면 가독성이 좀 더 좋아집니다 :)
```
return temp
        .map(Integer::parseInt)
        .collect(Collectors.toList());
```

출처 : https://marcin-chwedczuk.github.io/java-streams-best-practices
> You should have at most one stream method call per line. 
> This will make stream operations like map, filter and  collect easily recognizable.
```

kunggom replied:
알겠습니다. 다른 부분에는 적용했는데, 여기서는 실수로 빠트렸네요.

gracefulBrown commented on src/main/java/lotto/controller/ConsoleLottoGame.java:
```diff
@@ -0,0 +1,104 @@
+        return result;
+    }
+
+    private static List<Lotto> manualLottos(final int purchaseAmount) {
```
전반적으로 controller에 로직이 많은 것 같아요.
해당 메서드 같은 경우에는 일급컬렉션으로 묶어도 좋을 것 같네요 :)

kunggom replied:
콘트롤러 부분에 로직이 많기는 한데, MVC 패턴에 대한 경험 부족으로 어떤 식으로 나눠야 할지 잘 몰라서 일단 그대로 제출하였습니다. `manualLottos()` 메소드를 일급 컬렉션으로 묶는다는 것이 정확히 어떤 의미인지 잘 모르겠습니다. 수동으로 로또를 구매하는 과정만을 진행하는 클래스를 별도로 작성하라는 말씀이신가요?

gracefulBrown replied:
로직은 가급적이면 도메인으로 옮기는 방향으로 하고 controller에서는 흐름을 제어하는 정도의 역할만 가지면 좋을 것 같아요

일급컬렉션 언급은 이전 미션에서 이미 소개가 되었듯이 컬렉션을 랩핑한 별도의 클래스를 만드는것을 추천드리는 내용입니다.
로또 컬렉션을 별도의 (일급컬렉션)클래스로 분리할 경우 이와 관련된 로직은 해당 일급컬렉션으로 모아지면서
컨트롤러에서도 로직이 없어지는 효과를 가질 수 있을 것 같아요 :)

gracefulBrown commented on src/main/java/lotto/model/WinStat.java:
```diff
@@ -0,0 +1,78 @@
+    private static final String HORIZONTAL_BAR = "----------";
+    private static final String ITEM_DELIMITER = " - ";
+    private static final String QUANTITY = "개";
+    private static final String TOTAL_PROFIT_RATE = "총 수익률은 ";
```
출력 포맷 관련 로직은 도메인 보다는 뷰로 옮기는건 어떨까요?

kunggom replied:
객체 자체를 뷰에서 받아서 출력 스트링을 생성하면 될까요?

gracefulBrown commented on src/main/java/lotto/model/Rank.java:
```diff
@@ -0,0 +1,77 @@
+    FOURTH(4, 50_000), // 4등
+    THIRD(5, 1_500_000), // 3등
+    SECOND(5, 30_000_000, ", 보너스 볼 일치"), // 2등
+    FIRST(6, 2_000_000_000); // 1등;
```
언더바 이용 숫자 표현 👍

kunggom replied:
이게 은근히 가독성에 도움이 되더라고요.

gracefulBrown commented on src/main/java/lotto/controller/ConsoleLottoGame.java:
```diff
@@ -0,0 +1,104 @@
+        return lottos;
+    }
+
+    private static WinningLotto getWinningLotto() {
```
WinningLotto 생성 로직인것 같은데 WinningLotto로 옮기는건 어떨까요?

kunggom replied:
입출력 관련 로직이 있어서 일부러 모델이 아니라 컨트롤러 로직으로 넣었는데, 입출력을 모델에 넣는 것은 별로 좋은 생각이 아닌 것 같더라고요. 혹시 더 좋은 방법이 있을까요?

gracefulBrown replied:
제가 보기엔 입출력 관련 로직이라기 보다는 생성자의 역할을 하는 메서드로 보였습니다ㅎㅎ
만약 생성자로서의 역할이 아니라면 입력받은 숫자를 통해 WinningLotto를 만들어주는 로직을 View에서 처리한 후 controller로 넘겨주는건 어떨까요? 한번 고민해보세요 :)

gracefulBrown reviewed (changes requested):
커밋이 하나로 뭉쳐서 올라왔네요??
의도된것인지는 모르겠지만 작업별 분리하여 커밋해주면 좋을 것 같아요 :)
전체적으로 가독성을 신경을 많이 써서 잘 구현하셨습니다 👍 
피드백 남겼으니 확인해주시고요 :) 
피드백 내용으로 다른 부분에서도 같이 확인해보시면 좋을 것 같습니다.
궁금한점 있으면 언제든 DM 보내주세요~

kunggom commented:
커밋이 하나로 뭉친 것은 제 실수입니다. 나머지 사항은 답변 받는대로 적용하도록 하겠습니다.

kunggom committed:
수정: 피드백 반영

입출력 쪽의 이름을 바꾸고, 입출력과 관계된 코드를 모두 view 쪽으로 옮겼다.
처음에는 안 좋지 않나 생각했지만, 일단 이대로 가볼 생각이다.
그나저나 정신없이 하다 보니 커밋 자주 하는 게 잘 안 된다.

gracefulBrown commented on src/main/java/lotto/model/LottoRule.java:
```diff
@@ -0,0 +1,75 @@
+                + "개로 이루어져 있어야 합니다.";
+    }
+
+    //    public int getNumberStart() {
```
git을 믿고 현 시점에서 사용하지 않는 코드는 과감히 삭제해주세요ㅎㅎ

kunggom replied:
아직 코드 삭제 같은 부분은 완전히 과감해지기가 어렵네요 ㅜㅜ 노력하겠습니다.

gracefulBrown commented on src/main/java/lotto/view/ConsoleInput.java:
```diff
@@ -0,0 +1,132 @@
+        if (!message.equals(EMPTY)) {
+            System.out.println(message);
+        }
+        return SCANNER.nextLine().strip();
```
컴파일 에러가 나는것 같아요 ;;

kunggom replied:
제 환경에서는 아무 문제가 없습니다. 어쩌면 자바 버전 문제일지도 모르겠네요.

gracefulBrown replied:
아 그럴 수 있겠네요 11을 사용 중이군요 :)

gracefulBrown commented on src/main/java/lotto/controller/ConsoleLottoGame.java:
```diff
@@ -0,0 +1,45 @@
+        MAKER = new RandomLottoMaker(RULE);
+    }
+
+    private static int getAutoPurchaseCount(final int AllPurchaseCount, final int manualPurchaseCount) {
```
수동으로 구매 할 로또의 갯수를 검증하는 로직인가요?
만약 갯수(원시값)를 랩핑한 클래스가 있다면 도메인으로 빠질 수 있는 로직 같은데 한번 시도해보시겠어요?

원시값을 랩핑하면 생기는 일 : [링크](https://developerfarm.wordpress.com/2012/01/27/object_calisthenics_4/)

kunggom replied:
원시값을 어떻게 포장해야 할지 감이 잘 안 옵니다. `enum`은 적합하지 않아 보이고… 혹시 샘플 코드가 있나요?

gracefulBrown commented on src/main/java/lotto/view/ConsoleInput.java:
```diff
@@ -0,0 +1,132 @@
+        return result;
+    }
+
+    public List<Lotto> manualLottos(final int purchaseAmount) {
```
Input에서 도메인을 생성하는 로직이 마음에 들지 않는다면
Input에서는 원시값(List<String>)을 추출해서 Controller로 보내고
Controller에서 도메인에 접근해 생성하도록 하면 좋을 것 같네요 :)

https://github.com/woowacourse/java-lotto/pull/53/files#diff-d473d886c5835ec8f8ace178a8cbf7c6R116 도 마찬가지 입니다 :)

kunggom replied:
저도 그리 생각했는데, 입력값의 검증 부분이 걸리더라고요. 저는 입력값의 최종 검증 또한 로또 클래스의 책임이라고 생각했거든요. 만약에 입력값의 검증 부분을 분리한다면, 검증되지 않은 값이 들어간 로또의 인스턴스가 생성될 위험이 있다고 판단했습니다. 그런데 리스트를 대상으로 한꺼번에 일괄적으로 값 검증을 한다면, 만약에 사용자가 입력 중에 1번만 오타를 내어도 모든 입력을 버리고 새로 입력을 시작해야 하는 사태가 벌어집니다. 그렇다고 로또 객체와는 별개로 중복된 입력값 검증 코드를 작성하는 것은 DRY 원칙에 어긋나는 것 같고 해서, 일단은 이렇게 구현했습니다.

gracefulBrown replied:
> 만약에 입력값의 검증 부분을 분리한다면, 검증되지 않은 값이 들어간 로또의 인스턴스가 생성될 위험이 있다고 판단했습니다. 

**입력**에 대한 검증은 input에서 하고 도메인을 생성하는데 필요한 **입력값**에대한 검증은 도메인에서 할 수 있지 않을까요?

> 한꺼번에 일괄적으로 값 검증을 한다면, 만약에 사용자가 입력 중에 1번만 오타를 내어도 모든 입력을 버리고 새로 입력을 시작해야 하는 사태가 벌어집니다.

Input에서 List<Lotto>의 번호를 받는게 아니라 한번의 Lotto를 받은 후 controller에서 반복문을 돌면서 로또를 각각 만들면 어떨까요?

gracefulBrown commented on src/main/java/lotto/model/RandomLottoMaker.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    private List<Integer> makeList() {
+        List<Integer> temp = new ArrayList<>();
```
변수의 네이밍 시 의도를 분명히 밝히면 좋을 것 같습니다.
다른 코드에서도 temp나 의도를 분명히 밝히지 않은 네이밍을 확인해보세요 :)

참고 자료
> Clean Code > 2. Meaningful Names > Use Intention-Revealing Names

kunggom replied:
지역적으로 사용한 다음 버리는 변수에 굳이 작명을 고민할 필요가 있을까요? 예를 들어 `for`문에서 `i`나 `j`와 같은 변수명이 사용되는 것 또한 해당 변수들이 scope를 벗어나지 않기 때문에 별 문제가 되지 않기 때문인데, 이런 변수에 작명을 고민하는 것은 오히려 시간 낭비일 뿐이라고 생각합니다.

gracefulBrown replied:
저도 for문에서의 i나 j의 네이밍은 불필요할 수 있다고 생각합니다만
만약 for문안에서 index의 역할을 한다던지 특별한 일을 수행 할 경우에는 네이밍이 필요하다고 생각합니다.

아래의 메서드는 temp를 사용하는 다른 메서드인데요,
이와 같은 메서드에서의 temp, check 네이밍은 해당 메서드를 처음 접하는 개발자에게 아래와 같은 생각을 야기합니다.
- getInput(message)의 결과물이 무엇인지 알 수 없으며 [0]을 왜 사용하는지 유추할 수 없음
- isOnlyNumber의 파라미터로 사용되는 용도를 알 수 없으며 check변수가 무엇을 체크하는 지 유추할 수 없음
- temp = getInput(ERROR_INPUT_NUMBER)[0] 에서 재정의가 계속 되고 있음
```
    private static int getSingleInt(String message) {
        String temp = getInput(message)[0];
        boolean check = isOnlyNumber(temp);
        while (!check) { // 입력에 문제가 있으면 계속 재입력 요구
            temp = getInput(ERROR_INPUT_NUMBER)[0];
            check = isOnlyNumber(temp);
        }
        return Integer.parseInt(temp);
    }
```
https://github.com/woowacourse/java-lotto/pull/53/files#diff-d473d886c5835ec8f8ace178a8cbf7c6R67

정리하자면 
- scope를 벗어나지 않아 그 행위와 의도를 유추할 수 있는경우에는 큰곰의 의견에 동의합니다.
- 하지만 여러가지 일을 수행하는 메서드 내에서 특별한 일을 수행하거나 다른 메서드에서 사용하는 경우, 그 행위와 의도가 변하거나 유추하기 어렵다면 네이밍 고민이 필요하다고 생각합니다.

gracefulBrown commented on src/main/java/lotto/view/ConsoleOutput.java:
```diff
@@ -0,0 +1,95 @@
+            int value = entry.getValue();
+            String temp
+                    = key.equals(Rank.MISS)
+                    ? EMPTY // 낙첨은 당첨 통계에서 출력하지 않도록 함
```
주석을 사용하면 좋지 않다는 의미는 아니지만 
현재 메서드에는 여러가지 일을 수행하고 있기 때문에
메서드 분리 후 메서드 네이밍을 통해 드러내는 것이 좋을 수 있습니다.

kunggom replied:
좋은 조언 감사합니다. 메소드를 분리하도록 하곘습니다. 하지만 저는 메소드를 분리한 다음에도 이 정도 주석은 달아주는 것이 좋다고 생각합니다.

gracefulBrown commented on src/main/java/lotto/model/Rank.java:
```diff
@@ -0,0 +1,77 @@
+
+    private static final int WINNING_MIN_COUNT = 3;
+    private static final String NOT_VALID_VALUE = "는 유효하지 않은 값입니다.";
+    private static final String FAILED = "낙첨";
```
출력 관련 로직이 도메인에 남아있는 것 같습니다.

kunggom replied:
지적 감사합니다. 해당 부분을 작성하고 시간이 지나다 보니까 완전히 까먹었네요. `ConsoleOutput`으로 분리하도록 하겠습니다.

gracefulBrown reviewed (approved):
피드백 반영하시느라 고생하셨어요~
커밋 나누기가 잘 이행되지 않는다면 README에 작성한 것 처럼 
피드백 반영도 수행 할 작업의 목록을 하나씩 제거하면서 진행해보세요 :)
몇가지 피드백을 추가로 남겼으니 웹 미션 수행하면서 참고하면서 진행해보세요 ^^
